### PR TITLE
Enqueue period_bytes during playback

### PR DESCRIPTION
**Pull request by MatiasVara**
Enqueue period_bytes during playback.

### CODE REVIEW

**Comment by slp** [comment]
In which scenarios is `Wait a period to return a request` needed?

In my tests, ensuring that we feed `period_bytes` to pipewire on `process()` (something that `Enqueue period_bytes during playback` does) should also imply that the latter is only called when new data is needed, which is also when those buffers should be consumed and the guest notified.

File: crates/sound/src/device.rs [hunk 1]
```diff
@@ -489,7 +467,7 @@ impl VhostUserSoundThread {
             }
         }
 
-        Ok(false)
+        Ok(true)
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs]: Why this change is needed?

IIUC `handle_event()` should return `true` only if an exit event is received, but we are discussing removing that return value: https://github.com/rust-vmm/vhost/issues/144

I think we should check better that part into the code. It looks like we are not following that requirement.

**Comment by MatiasVara** [comment]
> In which scenarios is `Wait a period to return a request` needed?
> 
> In my tests, ensuring that we feed `period_bytes` to pipewire on `process()` (something that `Enqueue period_bytes during playback` does) should also imply that the latter is only called when new data is needed, which is also when those buffers should be consumed and the guest notified.

I am not really sure in which scenarios this is required. In my tests, I did not see the need either. For example, `aplay FrontLeft.wav` is ~12 periods and I measured:
```
going to sleep next_buffer: 125ms, elapsed: 180.417394ms
going to sleep next_buffer: 250ms, elapsed: 308.361508ms
going to sleep next_buffer: 375ms, elapsed: 436.354594ms
going to sleep next_buffer: 500ms, elapsed: 564.268021ms
going to sleep next_buffer: 625ms, elapsed: 670.913404ms
going to sleep next_buffer: 750ms, elapsed: 798.736913ms
going to sleep next_buffer: 875ms, elapsed: 926.872216ms
going to sleep next_buffer: 1s, elapsed: 1.054848565s
going to sleep next_buffer: 1.125s, elapsed: 1.161514281s
going to sleep next_buffer: 1.25s, elapsed: 1.289499503s
going to sleep next_buffer: 1.375s, elapsed: 1.417477883s
going to sleep next_buffer: 1.5s, elapsed: 1.545446985s   
```
You can see the dropping is happening after the theoretical next_buffer. 
Crosvm does it as in this PR but the audio backend is not pw. That made me think that this mechanism may be required in some cases. I did it for pw but I think this shall be independent of the audio backend. I am OK to drop this commit and just keep it in case we face this issue in the future for a different audio backend.

File: crates/sound/src/device.rs (hunk 1 above)
Comment (MatiasVara) [reply crates/sound/src/device.rs]: I did not meant to change this. I will fix it. Thanks.

Comment (MatiasVara) [reply crates/sound/src/device.rs]: What do you mean with `I think we should check better that part into the code. It looks like we are not following that requirement.`?

**Comment by MatiasVara** [comment]
I add @epilys and @stsquad to the discussion.

File: crates/sound/src/device.rs (hunk 1 above)
Comment (stefano-garzarella) [reply crates/sound/src/device.rs]: Yep, but double-checking it seems we don't use that return value... so if we don't need, maybe we can just return `()`

Comment (MatiasVara) [reply crates/sound/src/device.rs]: We are not currently using it and `handle_event()` is always returning `False`. How the returning value of `process_tx()` shall be used? What is an exit event?

Comment (stefano-garzarella) [reply crates/sound/src/device.rs]: I don't know how it can be used, honestly ;-) I think we can just remove it if it is not used and re-introduce when we need it. (maybe the idea was to return `false` if we don't have any more work to do? sorry, I don't remember)

About the `exit_event` is the eventfd we use to tell the event loop we want to exit: https://github.com/virtio-sound/vhost-device/blob/cbc80bee229ed0b827d45ef14d3b89d9396d72f8/crates/sound/src/device.rs#L569

It seems now it is directly handled by the backend loop in vhost-user-backend, so the event should not be propagated to our `handle_event`.
Maybe it was an old design. I think we will remove the return value. I'll look into it next week (for the vhost-user-backend point of view).

**Comment by epilys** [comment]
Hi all,

The period is over when the audio backend has played `period_bytes` which we control since we are the medium between the guest and the host audio backend. I don't think we should rely on a system clock for that, monotonic or otherwise.

A change I've made in the qemu virtio-snd device is only reading bytes from the IO message when the host audio backend is ready to play them. I think we should keep `Buffer` as it is but make the bytes reading lazy. I can make this change in `vhost-device` as well if you agree with it.

**Comment by slp** [comment]
> A change I've made in the qemu virtio-snd device is only reading bytes from the IO message when the host audio backend is ready to play them. I think we should keep `Buffer` as it is but make the bytes reading lazy. I can make this change in `vhost-device` as well if you agree with it.

That's what the first two commits of this PR do for the pipewire backend. I think the third can be dropped.

File: crates/sound/src/device.rs (hunk 1 above)
Comment (MatiasVara) [reply crates/sound/src/device.rs]: I see, I'll address the returning value for those functions in a different PR. I just created #39 to not forget. Thanks.

**Comment by MatiasVara** [comment]
I've dropped the third commit and renamed the PR.

**Comment by epilys** [comment]
@MatiasVara would you mind waiting to merge this until I have my alternative proposal ready?

**Comment by stefano-garzarella** [comment]
@epilys it would be nice though to have a review on this and also on #32 to see what could be improved

**Comment by epilys** [comment]
@stefano-garzarella most definitely, I'm trying to keep up with other work stuff lately and I am at times overwhelmed, but I will try to give reviews!

**Comment by stefano-garzarella** [comment]
@epilys sure, take your time ;-)
for us will be very helpful to understand what to improve!

File: crates/sound/src/stream.rs [hunk 2]
```diff
@@ -251,33 +252,66 @@ impl std::fmt::Debug for Buffer {
...
+        let mut buf_pos = 0;
+
+        while count > 0 {
+            // payload finishes at index (descriptor.len() - 2)
+            if self.idx_curr_desc > descriptors.len() - 1 {
+                panic!(
+                    "idx_desc: {} > len: {}!",
+                    self.idx_curr_desc,
+                    descriptors.len() - 1
+                );
```
Comment (dorindabassey) [inline crates/sound/src/stream.rs:279]: I think we can avoid the need to panic here, and instead return an error what do you think.

Comment (MatiasVara) [reply crates/sound/src/stream.rs:279]: I think you're right. I was not sure what to do in this situation. The panic is because consume() is reading more that it is possible. This should never happen because caller is passing to consume() a buf with the correct size, e.g., request.len() - request.pos(). However, this may not be always the case and consume() may read less than the buffer size. I'll avoid panic and instead, log a msg and return.

**Comment by epilys** [comment]
@MatiasVara check out #42. I'm rewriting the polling/buffer dropping synchronization in alsa to make it as idiomatic (for alsa) as possible.

File: crates/sound/src/audio_backends/pipewire.rs [hunk 3]
```diff
@@ -341,15 +339,18 @@ impl AudioBackend for PwBackend {
...
-
-                                buffer.pos = start;
-
-                                if start >= buffer.data_descriptor.len() as usize {
+                                // consume() always tries to read p.len() unless error
+                                let n_bytes = match buffer.consume(p) {
+                                    Ok(n_bytes) => n_bytes as usize,
+                                    Err(_) => {
+                                        log::error!("Failed to read from guest memory");
+                                        0
```
Comment (MatiasVara) [inline crates/sound/src/audio_backends/pipewire.rs]: @dorindabassey do you think it is better to set `n_bytes` to `0` or just return if an error occurs?

Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: yea, I think it's ok to log the error and set n_bytes to 0

**Review by epilys — COMMENTED** [review]
This is solved in #42 with minimal changes to current code, for example the diff in `pipewire.rs` was only one line in that commit. Plus checking chain indices, lengths is very manual && error prone, while correctness is enforced statically in #42. Also, since a request is split into separate buffers per descriptor, only a small part of the guest's data is read, giving more time to the VM to populate the bytes that follow, while the `Request` type here reads all the descriptors at once. Making this change loses the granularity of the previous logic.

**Comment by slp** [comment]
> Also, since a request is split into separate buffers per descriptor, only a small part of the guest's data is read, giving more time to the VM to populate the bytes that follow, while the Request type here reads all the descriptors at once. Making this change loses the granularity of the previous logic.

The specs clearly state that the enqueued buffers must be of the size `period_bytes`:

```
Since all buffers in the queue (with one exception) should be of the size period_bytes, the completion of
such an I/O request can be considered an elapsed period notification
```

Given that the buffer is already available and the worker thread is properly synchronized with the audio stream (which is the case, at least for the pipewire backend), how is *not* reading `period_bytes` each time the thread wakes up a good thing? It just forces the device to wake up more frequently, wasting CPU cycles.

**Comment by epilys** [comment]
the `Buffer` type is not the same thing as the `buffer` in the spec, you're confusing those two. I'm talking about the content of each descriptor in the chain.

**Comment by stefano-garzarella** [comment]
LGTM. The `process_tx()` now looks like what we do in other devices as well, and it's easier for me to follow. It's fine for me to merge.

@ablu @vireshk @stsquad  what do you think?

**Comment by stefano-garzarella** [comment]
> the `Buffer` type is not the same thing as the `buffer` in the spec, you're confusing those two. I'm talking about the content of each descriptor in the chain.

Can you list the differences?

**Review by stefano-garzarella — APPROVED** [review]

File: crates/sound/src/stream.rs [hunk 4]
```diff
@@ -251,33 +252,67 @@ impl std::fmt::Debug for Buffer {
...
-            .desc_chain
-            .memory()
-            .read(
-                buf,
-                addr.checked_add(offset)
-                    .expect("invalid guest memory address"),
-            )
-            .map_err(|_| Error::DescriptorReadFailed)?;
-        Ok(len as u32)
+    pub fn consume(&mut self, buf: &mut [u8]) -> Result<u32> {
```
Comment (Ablu) [inline crates/sound/src/stream.rs:267]: Looks like everyone has to work around that something like https://github.com/rust-vmm/vm-virtio/pull/33 is missing. It looks like every device falls into two categories. Either it implements its own descriptor chain walking or ends up glossing over the fact that descriptors may be sliced up arbitrarily. For SCSI we already carry descriptor chain walking. This would be another variant.

I think we should just provide it as part of vm-virtio and remove our own attempts of doing it from vhost-device.

File: crates/sound/src/device.rs [hunk 5]
```diff
@@ -388,100 +388,76 @@ impl VhostUserSoundThread {
...
+            }
+
+            let status_desc = descriptors[descriptors.len() - 1];
+            if status_desc.len() as usize != size_of::<VirtioSoundPcmStatus>() {
+                return Err(Error::UnexpectedDescriptorSize(
+                    size_of::<VirtioSoundPcmStatus>(),
+                    status_desc.len(),
+                )
+                .into());
+            }
```
Comment (Ablu) [inline crates/sound/src/device.rs:445]: Code just moved around, but: The spec probably allows all these structs to be broken up across multiple descriptors?

File: crates/sound/src/audio_backends/alsa.rs [hunk 6]
```diff
@@ -145,17 +145,18 @@ fn write_samples_direct(
...
             return Ok(false);
         };
-        let mut buf = vec![0; buffer.data_descriptor.len() as usize];
-        let read_bytes = buffer
-            .consume(&mut buf)
-            .expect("failed to read buffer from guest");
+        let mut buf = vec![0; buffer.len];
+        let read_bytes = buffer.consume(&mut buf).unwrap_or_else(|_| {
+            log::error!("Failed to read from guest memory");
+            0
```
Comment (Ablu) [inline crates/sound/src/audio_backends/alsa.rs:151]: This .unwrap + log but continue seems to be a common pattern here? Is there a reason why we can't just error out of the current function (and then, maybe log it at a higher level from some central place)?

**Review by Ablu — COMMENTED** [review]
I admit that I miss context to properly follow this. There seems to be a lot of moving code around that I fail to follow. Maybe some more detailed explanation in the commit message (or if possible a breakup across multiple commits) would help.

**Comment by epilys** [comment]
> > the `Buffer` type is not the same thing as the `buffer` in the spec, you're confusing those two. I'm talking about the content of each descriptor in the chain.
> 
> Can you list the differences?

The buffer in the spec refers to the part of the I/O message that carries the audio data, regardless of how it's split into descriptors etc. In this crate the I/O message has its own type, and because the audio data part of it might be broken up into multiple descriptors (read-only for TX and write-only for RX), each descriptor subfragment is represented by `Buffer`. This way we can associate each fragment with its own descriptor in the chain.

> The process_tx() now looks like what we do in other devices as well, and it's easier for me to follow.

It looks like a gratuitous change to me. The commit is titled "Enqueue period_bytes during playback" and it changes all the TX process message code which is unrelated.

In my PR, this is done in a few lines of code.

**Comment by slp** [comment]
> The buffer in the spec refers to the part of the I/O message that carries the audio data, regardless of how it's split into descriptors etc. In this crate the I/O message has its own type, and because the audio data part of it might be broken up into multiple descriptors (read-only for TX and write-only for RX), each descriptor subfragment is represented by Buffer. This way we can associate each fragment with its own descriptor in the chain.

The driver is required to always enqueue a buffer of a size of `period_bytes` (and I've empirically tested this is the case for the Linux driver). How this buffer is split in different descriptors in a descriptor chain depends on the configuration of the virtqueue and device implementations should *not* care about it. Which is precisely the reason why many (most?) Rust device implement abstractions to consume a whole descriptor chain as if it were a single linear buffer, just as this PR does.

IMO doing this for virtio-snd makes the code cleaner and easier to read, more aligned with other device implementations, and more efficient (the thread in the backend doesn't need to wake up so often).

**Comment by MatiasVara** [comment]
> Also, since a request is split into separate buffers per descriptor, only a small part of the guest's data is read, giving more time to the VM to populate the bytes that follow, while the `Request` type here reads all the descriptors at once. Making this change loses the granularity of the previous logic.

It is sufficient to wait one period in order to be able to read the entire period_bytes. `Aplay`, for instance, reads the fifth period only after four periods. The problem arises when the device reads _immediately_ once a period_bytes arrives in the available ring. After four periods, which is the number of pre-buffered buffers,. e.g., between PREPARE and START, it looks like enough time in the experiments. Generally speaking, the pre-buffering before START gives enough delay so the pipewire backend can read a period in the right time. Also, in the case of pipewire, this allows us to use the full pw buffer that is 32kb.

> the Buffer type is not the same thing as the buffer in the spec, you're confusing those two. I'm talking about the content of each descriptor in the chain.

I've already pointed this months ago at https://lists.oasis-open.org/archives/virtio-comment/202308/msg00131.html.

> It looks like a gratuitous change to me. The commit is titled "Enqueue period_bytes during playback" and it changes all the TX process message code which is unrelated.

I will split the PR into two commits. Thanks.

**Comment by epilys** [comment]
> The driver is required to always enqueue a buffer of a size of `period_bytes` (and I've empirically tested this is the case for the Linux driver). How this buffer is split in different descriptors in a descriptor chain depends on the configuration of the virtqueue and device implementations should _not_ care about it. Which is precisely the reason why many (most?) Rust device implement abstractions to consume a whole descriptor chain as if it were a single linear buffer, just as this PR does.
> 
> IMO doing this for virtio-snd makes the code cleaner and easier to read, more aligned with other device implementations, and more efficient (the thread in the backend doesn't need to wake up so often).

Doesn't seem like it?

https://github.com/virtio-sound/vhost-device/pull/35/files#diff-765431f3d066484dea5f0be9c6b049982e8fdfe56ee5c3443636e8355c771056R268-R279

The audio backend thread does not wake up more often. In my PR the thread consumes as many `Buffer`s are available while there is still available bytes in the host audio buffer to populate.

Honestly, none of the arguments here seem genuine to me and I disagree with merging this PR.

File: crates/sound/src/stream.rs (hunk 4 above)
Comment (slp) [reply crates/sound/src/stream.rs:267]: That'd be nice. I suggest importing the descriptor_utils from crosvm, which is quite powerful and flexible.

**Comment by MatiasVara** [comment]
The pw thread loop wakes up less often when using the whole buffer we get in the `process()` callback. This happens because the pw engine runs out of data slower. I think we should also do this optimization after #42, so once #42 is merged, I am going to split this PR and rebase on top of #42.

**Comment by slp** [comment]
> The audio backend thread does not wake up more often. In my PR the thread consumes as many Buffers are available while there is still available bytes in the host audio buffer to populate.

It certainly is for the pipewire backend, as `process()` is called on demand, which implies the more bytes you feed it, the less often it calls you.

While I haven't checked, I suspect that what happens with the alsa backend is that it's already reading `period_bytes` by always processing all available buffers. Have you checked how many bytes are fed each time the thread wakes up?

> Honestly, none of the arguments here seem genuine to me and I disagree with merging this PR.

Okay, let's take a step back and split things up to discuss them individually, because there are at least 4 different things on the table:

1. The rewrite of `process_tx`. While I *personally* like @MatiasVara approach, I do acknowledge that it's not strictly necessary and that it's more a matter of tastes than anything else.
2. The implementation of an abstraction for easily consuming each descriptor chain as if it were a linear buffer. I *do* think this is an important one that should make everyone's life easier without hurting any backend implementation.
3. Delay reading the descriptors from `process_tx` to when they're actually needed. This *must* be implemented for the device to operate properly. It's achieved, through slightly different approaches, by both #42 and this one.
4. Ensure pipewire is properly fed on each call to `process()`. It's easier and more elegant with an abstraction over the descriptor chain (as long a solution for `3` is implemented), but can be done without it.

If I can suggest something, I would merge #42 (fixing `3`), implement a descriptor chain consumer abstraction on top of it (fixing `2`), adapt both alsa and pipewire backends to use it (fixing `4`), and drop `1`.

What do you think? @epilys @MatiasVara @stefano-garzarella @dorindabassey

**Comment by stefano-garzarella** [comment]
> If I can suggest something, I would merge #42 (fixing `3`), implement a descriptor chain consumer abstraction on top of it (fixing `2`), adapt both alsa and pipewire backends to use it (fixing `4`), and drop `1`.
> 
> What do you think? @epilys @MatiasVara @stefano-garzarella @dorindabassey

@slp I totally agree with this plan!

**Comment by dorindabassey** [comment]
I agree with your suggestions @slp 
my only concern is regarding the rewrite of the `process_tx` the way it was handled in this PR enforces readability and clarity (probably to be done in another PR).

**Comment by epilys** [comment]
1. ()
2. The handling of descriptor chains is a general vhost-device issue (in specific see rust-vmm/vm-virtio, virtio-queue crate), but in this specific case buffers are split, we can't treat them as linear. Correct me if I'm wrong, in case there's an endianness mismatch between host and guest, the I/O message bytes will need to be byte swapped for example (not yet addressed anywhere yet, at least by myself) so a linear buffer becomes impossible in that case. PCM is sample-granular, not byte-granular. Unless all sample sizes and interleaved channel combinations happen to align perfectly - if I understand it correctly, DMA/MMAPing buffers in PCM is an optimization for hardware and not the general case.

    What's wrong with using the byte slices in the same granularity they are given to us? Why make an extra abstraction? Is there something that is not working right now that this would fix?
3. yes,
4. See https://github.com/virtio-sound/vhost-device/pull/42/commits/af77b0429147c7e1c8772a7f77b551a890384733#diff-8fc1266fb79d2060429a944e656b12d4a4fb3c8f5e6d9dcce0f4fb1e63d00386

In short, I don't believe there's a compelling reason not to merge #42 once review comments/issues are addressed. It'd be really unfortunate for me to delay merging and upstreaming the alsa backend if it's fundamentally functional. No hurrying in doing so of course, lest there be problems and issues we would miss.

**Comment by stefano-garzarella** [comment]
> 1. ()
> 
>     2. The handling of descriptor chains is a general vhost-device issue (in specific see rust-vmm/vm-virtio, virtio-queue crate), but in this specific case buffers are split, we can't treat them as linear. Correct me if I'm wrong, in case there's an endianness mismatch between host and guest, the I/O message bytes will need to be byte swapped for example (not yet addressed anywhere yet, at least by myself) so a linear buffer becomes impossible in that case.

For virtio spec point of view, a descriptors chain should be threat as a linear buffer. It's like an `iovec` used in `sendmsg(2)`.

The device or driver must reassemble it as if it were a single buffer. In fact, the device can't make any assumptions about how many descriptors are in a chain.

A practical example is [this bug in the rust-vmm vsock device](https://github.com/rust-vmm/vm-virtio/issues/204) that was there because the device always assumed 2 descriptors (header, payload), while the driver was optimized to use only one.

In fact in the specification it is not defined how the buffer should be divided into multiple descriptors, because it could potentially be even 1 descriptor per byte or a single descriptor for the whole buffer.

**Comment by epilys** [comment]
@stefano-garzarella Exactly. This is why the linear buffer mentions here is confusing me, the actual data we get can be scattered over multiple descriptors. Unless I've misunderstood what the "as if [..] linear buffer" argument was referring to.

**Comment by stefano-garzarella** [comment]
Closing, next version will be opened in https://github.com/rust-vmm/vhost-device repo
